File: zeta-client-cpp/src/nativeMain/kotlin/de/gematik/zeta/Hello.kt
/*
 * #%L
 * ZETA-Client
 * %%
 * (C) EY Strategy & Transactions GmbH, 2025, licensed for gematik GmbH
 * %%
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 *
 * ******
 *
 * For additional notes and disclaimer from gematik and in case of changes by gematik find details in the "Readme" file.
 * #L%
 */

@file:OptIn(ExperimentalNativeApi::class, ExperimentalForeignApi::class)

import de.gematik.zeta.sdk.BuildConfig
import de.gematik.zeta.sdk.StorageConfig
import de.gematik.zeta.sdk.TpmConfig
import de.gematik.zeta.sdk.ZetaSdk
import de.gematik.zeta.sdk.ZetaSdkClient
import de.gematik.zeta.sdk.attestation.model.PlatformProductId
import de.gematik.zeta.sdk.authentication.AuthConfig
import de.gematik.zeta.sdk.authentication.smb.SmbTokenProvider
import de.gematik.zeta.sdk.authentication.smcb.SmcbTokenProvider
import de.gematik.zeta.sdk.network.http.client.HttpClientExtension.httpGet
import de.gematik.zeta.sdk.network.http.client.HttpClientExtension.httpPost
import de.gematik.zeta.sdk.network.http.client.ZetaHttpClient
import de.gematik.zeta.sdk.network.http.client.ZetaHttpClientBuilder
import de.gematik.zeta.sdk.storage.InMemoryStorage
import interop.ZetaSdk_BuildConfig
import interop.ZetaSdk_Client
import interop.ZetaSdk_HttpClient
import interop.ZetaSdk_HttpHeader
import interop.ZetaSdk_HttpRequest
import interop.ZetaSdk_HttpResponse
import io.ktor.client.plugins.logging.LogLevel
import io.ktor.client.plugins.logging.Logger
import kotlinx.cinterop.ByteVar
import kotlinx.cinterop.CPointer
import kotlinx.cinterop.CPointerVar
import kotlinx.cinterop.ExperimentalForeignApi
import kotlinx.cinterop.StableRef
import kotlinx.cinterop.alloc
import kotlinx.cinterop.allocArray
import kotlinx.cinterop.asStableRef
import kotlinx.cinterop.get
import kotlinx.cinterop.nativeHeap
import kotlinx.cinterop.pointed
import kotlinx.cinterop.ptr
import kotlinx.cinterop.toKString
import platform.posix.free
import platform.posix.strdup
import kotlin.experimental.ExperimentalNativeApi

@CName(externName = "ZetaSdk_buildZetaClient")
fun ZetaSdk_buildSdkClient(
    buildConfig: CPointer<ZetaSdk_BuildConfig>,
): CPointer<ZetaSdk_Client> {
    val cBuildConfig = buildConfig.pointed
    val cStorageConfig = cBuildConfig.storageConfig!!.pointed
    val cTpmConfig = cBuildConfig.tpmConfig!!.pointed
    val cAuthConfig = cBuildConfig.authConfig!!.pointed
    val cSmbConfig = cAuthConfig.smbConfig?.pointed
    val cSmcbConfig = cAuthConfig.smcbConfig?.pointed

    val storageConfig = cStorageConfig.let { storageConfig ->
        StorageConfig(InMemoryStorage())
    }
    val tpmConfig = cTpmConfig.let { tpmConfig ->
        object : TpmConfig {}
    }
    val authConfig = cAuthConfig.let { authConfig ->
        val keystoreFile = cSmbConfig?.keystoreFile?.toKString() ?: ""
        val baseUrl = cSmcbConfig?.baseUrl?.toKString() ?: ""

        val subjectTokenProvider = when {
            keystoreFile.isNotEmpty() -> SmbTokenProvider(
                SmbTokenProvider.Credentials(
                    keystoreFile,
                    cSmbConfig?.alias?.toKString() ?: "",
                    cSmbConfig?.password?.toKString() ?: "",
                )
            )
            baseUrl.isNotEmpty() -> SmcbTokenProvider(
                SmcbTokenProvider.ConnectorConfig(
                    baseUrl,
                    cSmcbConfig?.mandantId?.toKString() ?: "",
                    cSmcbConfig?.clientSystemId?.toKString() ?: "",
                    cSmcbConfig?.workspaceId?.toKString() ?: "",
                    cSmcbConfig?.userId?.toKString() ?: "",
                    cSmcbConfig?.cardHandle?.toKString() ?: "",
                )
            )
            else -> throw Exception("Should specify SM-B / SMC-B subject token provider")
        }
        AuthConfig(
            authConfig.scopes?.toKList(authConfig.scopesCount)?.filterNotNull().orEmpty(),
            authConfig.exp.toLong(),
            authConfig.enableAslTracingHeader,
            subjectTokenProvider,
        )
    }
    val zetaSdkClient = cBuildConfig.let { buildConfig ->
        ZetaSdk.build(
            buildConfig.resource?.toKString() ?: "",
            BuildConfig(
                buildConfig.productId?.toKString() ?: "",
                buildConfig.productVersion?.toKString() ?: "",
                buildConfig.clientName?.toKString() ?: "",
                storageConfig!!,
                tpmConfig!!,
                authConfig!!,
                platformProductId = PlatformProductId.AppleProductId("apple", "macos", listOf("bundleX")),
                ZetaHttpClientBuilder().disableServerValidation(true).logging(LogLevel.ALL, object : Logger {
                    override fun log(message: String) {
                        println(message)
                    }
                }),
            )
        )
    }
    return nativeHeap.alloc<ZetaSdk_Client>().let { sdkClient ->
        sdkClient.zetaSdkClient = StableRef.create(zetaSdkClient).asCPointer()
        sdkClient.ptr
    }
}

@CName(externName = "ZetaSdk_clearZetaClient")
fun ZetaSdk_clearZetaClient(
    sdkClient: CPointer<ZetaSdk_Client>
) {
    sdkClient.pointed.let { sdkClient ->
        sdkClient.zetaSdkClient!!.asStableRef<ZetaSdkClient>().dispose()
    }
    nativeHeap.free(sdkClient.rawValue)
}

@CName(externName = "ZetaSdk_buildHttpClient")
fun ZetaSdk_buildHttpClient(
    sdkClient: CPointer<ZetaSdk_Client>
): CPointer<ZetaSdk_HttpClient> {
    val zetaSdkClient = sdkClient.pointed.zetaSdkClient!!.asStableRef<ZetaSdkClient>().get()
    val logger = object : Logger {
        override fun log(message: String) {
            println(message)
        }
    }
    val zetaHttpClient = zetaSdkClient.httpClient {
        logging(LogLevel.ALL, logger)
        disableServerValidation(true)
    }
    return nativeHeap.alloc<ZetaSdk_HttpClient>().let { httpClient ->
        httpClient.zetaHttpClient = StableRef.create(zetaHttpClient).asCPointer()
        httpClient.ptr
    }
}

@CName(externName = "ZetaSdk_clearHttpClient")
fun ZetaSdk_clearHttpClient(
    httpClient: CPointer<ZetaSdk_HttpClient>
) {
    httpClient.pointed.let { sdkClient ->
        sdkClient.zetaHttpClient!!.asStableRef<ZetaSdkClient>().dispose()
    }
    nativeHeap.free(httpClient.rawValue)
}

@CName(externName = "ZetaHttpClient_httpGet")
fun ZetaHttpClient_httpGet(
    httpClient: CPointer<ZetaSdk_HttpClient>,
    httpRequest: CPointer<ZetaSdk_HttpRequest>,
): CPointer<ZetaSdk_HttpResponse>? {
    val zetaHttpClient = httpClient.pointed.zetaHttpClient!!.asStableRef<ZetaHttpClient>().get()
    val zetaHttpRequest = httpRequest.pointed

    return try {
        val url = zetaHttpRequest.url?.toKString() ?: ""
        val headers = zetaHttpRequest.headers
            ?.toKList(zetaHttpRequest.headersCount)
            ?.associate {
                val key = it?.key?.toKString() ?: ""
                val value = it?.value?.toKString() ?: ""
                key to value
            } ?: emptyMap()
        val result = zetaHttpClient.delegate.httpGet(url, headers)
        val resultHeaders = nativeHeap.allocArray<ZetaSdk_HttpHeader>(result.headers.size)
        result.headers.entries.forEachIndexed { index, header ->
            resultHeaders[index].key = strdup(header.key)!!
            resultHeaders[index].value = strdup(header.value)!!
        }
        nativeHeap.alloc<ZetaSdk_HttpResponse>().let { httpResponse ->
            httpResponse.status = result.status
            httpResponse.body = strdup(result.body)!!
            httpResponse.headers = resultHeaders
            httpResponse.error = null
            httpResponse.ptr
        }
    } catch (e: Exception) {
        e.printStackTrace()
        nativeHeap.alloc<ZetaSdk_HttpResponse>().let { httpResponse ->
            httpResponse.body = null
            httpResponse.error = strdup(e.message)!!
            httpResponse.ptr
        }
    }
}

@CName(externName = "ZetaHttpClient_httpPost")
fun ZetaHttpClient_httpPost(
    httpClient: CPointer<ZetaSdk_HttpClient>,
    httpRequest: CPointer<ZetaSdk_HttpRequest>,
): CPointer<ZetaSdk_HttpResponse>? {
    val zetaHttpClient = httpClient.pointed.zetaHttpClient!!.asStableRef<ZetaHttpClient>().get()
    val zetaHttpRequest = httpRequest.pointed

    return try {
        val url = zetaHttpRequest.url?.toKString() ?: ""
        val body = zetaHttpRequest.body?.toKString() ?: ""
        val headers = zetaHttpRequest.headers
            ?.toKList(zetaHttpRequest.headersCount)
            ?.associate {
                val key = it?.key?.toKString() ?: ""
                val value = it?.value?.toKString() ?: ""
                key to value
            } ?: emptyMap()
        val result = zetaHttpClient.delegate.httpPost(url, body, headers)
        val resultHeaders = nativeHeap.allocArray<ZetaSdk_HttpHeader>(result.headers.size)
        result.headers.entries.forEachIndexed { index, header ->
            resultHeaders[index].key = strdup(header.key)!!
            resultHeaders[index].value = strdup(header.value)!!
        }
        nativeHeap.alloc<ZetaSdk_HttpResponse>().let { httpResponse ->
            httpResponse.status = result.status
            httpResponse.body = strdup(result.body)!!
            httpResponse.headers = resultHeaders
            httpResponse.error = null
            httpResponse.ptr
        }
    } catch (e: Exception) {
        e.printStackTrace()
        nativeHeap.alloc<ZetaSdk_HttpResponse>().let { httpResponse ->
            httpResponse.body = null
            httpResponse.error = strdup(e.message)!!
            httpResponse.ptr
        }
    }
}

@CName(externName = "ZetaHttpResponse_destroy")
fun ZetaHttpResponse_destroy(
    httpResponse: CPointer<ZetaSdk_HttpResponse>,
) {
    httpResponse.pointed.let { httpResponse ->
        httpResponse.body?.let { free(it)  }
        httpResponse.error?.let { free(it)  }
        httpResponse.headers?.toKList(httpResponse.headersCount)?.forEach {
            it?.key?.let { free(it) }
            it?.value?.let { free(it) }
        }
    }
    nativeHeap.free(httpResponse.rawValue)
}

fun CPointer<CPointerVar<ByteVar>>.toKList(count: Int): List<String?> {
    return List(count) { i -> this[i]?.toKString() }
}

fun CPointer<ZetaSdk_HttpHeader>.toKList(count: Int): List<ZetaSdk_HttpHeader?> {
    return List(count) { i -> this[i] }
}

File: zeta-client-cpp/src/main/cpp/hello.cpp
/*
 * #%L
 * ZETA-Client
 * %%
 * (C) EY Strategy & Transactions GmbH, 2025, licensed for gematik GmbH
 * %%
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 *
 * ******
 *
 * For additional notes and disclaimer from gematik and in case of changes by gematik find details in the "Readme" file.
 * #L%
 */

#include <cstdlib>
#include <cstdio>
#include <cstring>
#include <iostream>

#include "hello_api.h"
#include "ws_client_api.h"

#define ARRAY_SIZE(arr) (sizeof(arr) / sizeof((arr)[0]))
#define BUFFER_SIZE 1024

char* stompConnectFrame(const char* host) {
    char* format = "CONNECT\naccept-version:1.2\nhost:%s\n\n\00";
    char* buffer = new char[BUFFER_SIZE];
    int printedCount = snprintf(buffer, BUFFER_SIZE, format, host);
    if (printedCount <= 0) {
        std::cout << "Error formatting CONNECT frame: " << printedCount << "\n";
        std::cout.flush();
        return NULL;
    }
    return buffer;
}

char* stompSubscribeFrame(char* id, char* contextPath, char* destination) {
    char* format = "SUBSCRIBE\nid:%s\ndestination:%s%s\n\n\00";
    char* buffer = new char[BUFFER_SIZE];
    int printedCount = snprintf(buffer, BUFFER_SIZE, format, id, contextPath, destination);
    if (printedCount <= 0) {
        std::cout << "Error formatting SUBSCRIBE frame: " << printedCount << "\n";
        std::cout.flush();
        return NULL;
    }
    return buffer;
}

char* stompSendFrame(char* contextPath, char* destination, char* bodyJson) {
    char* format = "SEND\ndestination:%s%s\ncontent-type:application/json\n\n%s\00";
    char* buffer = new char[BUFFER_SIZE];
    int printedCount = snprintf(buffer, BUFFER_SIZE, format, contextPath, destination, bodyJson);
    if (printedCount <= 0) {
        std::cout << "Error formatting SEND frame: " << printedCount << "\n";
        std::cout.flush();
        return NULL;
    }
    return buffer;
}

void sendPrescriptionCommands(ZetaSdk_WSSession* wsSession, char* wsContextPath) {
    char* createBody =
            "\n{\n"
            "\"prescriptionId\": \"RX-2025-100123\",\n"
            "\"patientId\": \"PAT-123456\",\n"
            "\"practitionerId\": \"PRAC-98765\",\n"
            "\"medicationName\": \"Ibuprofen 400 mg\",\n"
            "\"dosage\": \"1\",\n"
            "\"issuedAt\": \"2025-09-22T10:30:00Z\",\n"
            "\"expiresAt\": \"2025-12-31T23:59:59Z\",\n"
            "\"status\": \"CREATED\"\n"
            "}\n\00";

    char* sendFrame1 = stompSendFrame(wsContextPath, "/app/erezept.create", createBody);
    int sendFrame1Size = strlen(sendFrame1) + 1;
    ZetaSdk_WSSession_sendBinary(wsSession, sendFrame1, sendFrame1Size);
    free(sendFrame1);

    char* sendFrame2 = stompSendFrame(wsContextPath, "/app/erezept.read.1", "{}");
    int sendFrame2Size = strlen(sendFrame2) + 1;
    ZetaSdk_WSSession_sendBinary(wsSession, sendFrame2, sendFrame2Size);
    free(sendFrame2);
}

void connectAndSubscribe(ZetaSdk_WSSession* wsSession, char* host, char* wsContextPath) {
    char* connectFrame = stompConnectFrame(host);
    int connectFrameSize = strlen(connectFrame) + 1;
    ZetaSdk_WSSession_sendBinary(wsSession, connectFrame, connectFrameSize);
    free(connectFrame);

    ZetaSdk_WSMessage* message = ZetaSdk_WSSession_receiveNext(wsSession);
    if (message->type == WS_TEXT) {
        std::cout << "CONNECTED frame: \"" << message->data.text.text << "\"\n";
        std::cout.flush();
    }

    char* subscribeFrame1 = stompSubscribeFrame("sub-1", wsContextPath, "/topic/erezept");
    int subscribeFrame1Size = strlen(subscribeFrame1) + 1;
    ZetaSdk_WSSession_sendBinary(wsSession, subscribeFrame1, subscribeFrame1Size);
    free(subscribeFrame1);

    char* subscribeFrame2 = stompSubscribeFrame("sub-2", wsContextPath, "/user/queue/erezept");
    int subscribeFrame2Size = strlen(subscribeFrame2) + 1;
    ZetaSdk_WSSession_sendBinary(wsSession, subscribeFrame2, subscribeFrame2Size);
    free(subscribeFrame2);
}

void receiveMessages(ZetaSdk_WSSession* wsSession) {
    for (int i = 0; i < 2; i++) {
        ZetaSdk_WSMessage* incoming = ZetaSdk_WSSession_receiveNext(wsSession);

        if (incoming->type == WS_CLOSE) {
            std::cout << "WebSocket closed.\n";
            std::cout.flush();
            break;
        } else if (incoming->type == WS_TEXT) {
            std::cout << "Received: " << incoming->data.text.text << "\n";
            std::cout.flush();
        } else if (incoming->type == WS_BINARY) {
            std::cout << "Received " << incoming->data.binary.size << " bytes (binary)\n";
            std::cout.flush();
        }

        ZetaSdk_WSMessage_destroy(incoming);
    }
}

void extractHost(char* url, char* host, size_t hostSize) {
    char* p = strstr(url, "://");
    if (p) p += 3;
    else p = url;

    char* end = strchr(p, '/');
    size_t len = end ? (size_t)(end - p) : strlen(p);

    if (len >= hostSize) len = hostSize - 1;

    strncpy(host, p, len);
    host[len] = '\0';
}

void runWSSessionSample(ZetaSdk_Client* zetaSdkClient, char* wsBaseUrl, char* wsServerContextPath, char* poppToken) {
    std::cout << "WSSession Sample: start\n";
    std::cout.flush();

    static char wsHost[BUFFER_SIZE];
    extractHost(wsBaseUrl, wsHost, BUFFER_SIZE);

    static char* wsContextPath = wsServerContextPath;
    void (*wsHandler)(ZetaSdk_WSSession *) = [](ZetaSdk_WSSession *wsSession) {
        std::cout << "WSSession Handler: start\n";
        std::cout.flush();

        connectAndSubscribe(wsSession, wsHost, wsContextPath);
        sendPrescriptionCommands(wsSession, wsContextPath);
        receiveMessages(wsSession);
        ZetaSdk_WSSession_close(wsSession);

        std::cout << "WSSession Handler: end\n";
        std::cout.flush();
    };

    ZetaSdk_HttpHeader httpHeaders[] = {
            {"PoPP", poppToken },
    };

    ZetaSdk_Client_ws(zetaSdkClient, wsBaseUrl, wsHandler, httpHeaders, ARRAY_SIZE(httpHeaders));

    std::cout << "WSSession Sample: end\n";
    std::cout.flush();
}

void runReceiptPost(ZetaSdk_HttpClient* zetaHttpClient, char* poppToken) {
    ZetaSdk_HttpHeader httpHeaders[] = {
            {"PoPP", poppToken },
    };
    ZetaSdk_HttpRequest httpPostRequest = {
            "api/erezept",
            "{\"prescriptionId\":\"RX-2025-000003\",\"patientId\":\"PAT-123456\",\"practitionerId\":\"PRAC-98765\",\"medicationName\":\"Ibuprofen 400 mg\",\"dosage\":\"1\",\"issuedAt\":\"2025-09-22T10:30:00Z\",\"expiresAt\":\"2025-12-31T23:59:59Z\",\"status\":\"CREATED\"}",
            httpHeaders,
            ARRAY_SIZE(httpHeaders)
    };

    ZetaSdk_HttpResponse* httpPostResponse = ZetaHttpClient_httpPost(zetaHttpClient, &httpPostRequest);

    if (httpPostResponse->error == NULL) {
        std::cout << "Response status: " << httpPostResponse->status << "\n";
        std::cout << "Response body: " << httpPostResponse->body << "\n";
        std::cout.flush();
    } else {
        std::cout << "Error during httpPost: " << httpPostResponse->error << "\n";
    }
}

void runHttpClientSample(ZetaSdk_HttpClient* zetaHttpClient, char* poppToken) {
    std::cout << "HttpClient Sample: start\n";
    std::cout.flush();

    ZetaSdk_HttpHeader httpHeaders[] = {
            {"PoPP", poppToken },
    };

    ZetaSdk_HttpRequest httpGetRequest = {
        "hellozeta",
        NULL,
        httpHeaders,
        ARRAY_SIZE(httpHeaders)
    };
    ZetaSdk_HttpResponse* httpGetResponse = ZetaHttpClient_httpGet(zetaHttpClient, &httpGetRequest);
    if (httpGetResponse->error == NULL) {
        std::cout << "Response status: " << httpGetResponse->status << "\n";
        std::cout << "Response body: " << httpGetResponse->body << "\n";
        std::cout.flush();
    } else {
        std::cout << "Error during httpGet: " << httpGetResponse->error << "\n";
    }
    ZetaHttpResponse_destroy(httpGetResponse);

    runReceiptPost(zetaHttpClient, poppToken);

    std::cout << "HttpClient Sample: end\n";
    std::cout.flush();
}

int main() {
    std::cout << "Hello from C++!\n";
    std::cout.flush();

    char* resource = std::getenv("RESOURCE_URL");
    char* keystoreFile = std::getenv("SMB_KEYSTORE_FILE");
    char* alias = std::getenv("SMB_KEYSTORE_ALIAS");
    char* password = std::getenv("SMB_KEYSTORE_PASSWORD");

    char* baseUrl = std::getenv("SMCB_BASE_URL");
    char* mandantId = std::getenv("SMCB_MANDANT_ID");
    char* clientSystemId = std::getenv("SMCB_CLIENT_SYSTEM_ID");
    char* workspaceId = std::getenv("SMCB_WORKSPACE_ID");
    char* userId = std::getenv("SMCB_USER_ID");
    char* cardHandle = std::getenv("SMCB_CARD_HANDLE");

    char* wsBaseUrl = std::getenv("WS_BASE_URL");
    char* wsContextPath = std::getenv("WS_SERVER_CONTEXT_PATH");
    char* poppToken = std::getenv("POPP_TOKEN");

    const char* envValue = std::getenv("ASL_PROD");
    bool aslProdEnv = false;
    if (envValue == nullptr) {
        aslProdEnv = true;
    } else {
        aslProdEnv = strcmp(envValue, "true") == 0;
    }

    char* productId = "demo-client";
    char* productVersion = "0.2.0";
    char* clientName = "sdk-client";
    char* scopes[] = {"zero:audience"};

    ZetaSdk_StorageConfig storageConfig = {};
    ZetaSdk_TpmConfig tpmConfig = {};
    ZetaSdk_SmbConfig smbConfig = {
            keystoreFile,
            alias,
            password
    };
    ZetaSdk_SmcbConfig smcbConfig = {
            baseUrl,
            mandantId,
            clientSystemId,
            workspaceId,
            userId,
            cardHandle
    };
    ZetaSdk_AuthConfig authConfig = {
            scopes,
            ARRAY_SIZE(scopes),
            30,
            aslProdEnv,
            &smbConfig,
            &smcbConfig
    };
    ZetaSdk_BuildConfig buildConfig = {
            resource,
            productId,
            productVersion,
            clientName,
            &storageConfig,
            &tpmConfig,
            &authConfig
    };

    ZetaSdk_Client* zetaSdkClient = ZetaSdk_buildZetaClient(&buildConfig);
    ZetaSdk_HttpClient* zetaHttpClient = ZetaSdk_buildHttpClient(zetaSdkClient);

    runHttpClientSample(zetaHttpClient, poppToken);
    runWSSessionSample(zetaSdkClient, wsBaseUrl, wsContextPath, poppToken);

    ZetaSdk_clearHttpClient(zetaHttpClient);
    ZetaSdk_clearZetaClient(zetaSdkClient);

    return 0;
}
